### PR TITLE
Const contexts for literal expressions

### LINKED ISSUE #10195: Syntax for hinting literal type inference
Now that we have so many literal types we more than ever need new syntax that would make their use natural. Please consider the following:

``` typescript
const value = (true); // true
const value = true; // boolean

const value = ('a'); // 'a'
const value = 'a'; // string

const value = (1); // 1
const value = 1; // number

const value = ['a', 1]; // (string | number)[]
const value = (['a', 1]) // [string, number]
const value = ([('a'), (1)]) // ['a', 1];
```

Problem:
- There is no way to get a literal type value without explicit type annotation.

Solution:
- Add new syntax (or shall we say some new semantics to the old mostly unused syntax)

Highlights:
- Composable.
- Works fine with the existing syntax, doesn't need a new one.
- Highly unlikely to be a breaking change.
- Doesn't affect JavaScript semantics at all, can be executed as written and will work as expected.

Shortcomings:
- Undiscoverable syntax.
- Semantic conflicts (breaking changes) although rare:
  - generated code
  - conditional expressions
  - unintended excessive syntax

Prior work:
- Inspired by #9217 which is closed in favor of this one.

### LINKED ISSUE #20195: Suggestion: a way to disable type widening in object literals
I understand the use of type widening with `let` and `var` variables (I use `const` ubiquitously so this is never a problem). But with object literals I pretty much _never_ want the fields to be widened, and yet that's the default behavior:

```ts
const o = { x: 3 }; // inferred type is { x: number }
```

The only way I know of to work around this is with casts or type annotations, or by passing the object literal immediately to a function expecting the narrower type, thereby guiding type inference (although if the literal is being returned from a function even that technique won't work).

This is an eternal headache and it crops up in so many places. Would it be possible to add a compiler flag that makes it so that object literals don't get widened? I'm happy to be responsible for adding type annotations in the extremely rare cases where I actually _want_ the widened type.

### Edit

It looks like compiler flags are generally frowned upon to solve this problem, so let me amend my suggestion to take account of the discussion below. @mhegazy says,

> We have talked in the past about a `readonly` modifier on property declarations.. e.g.
> ```ts
> const o = { readonly x: 3 }; 
> ```
> this allows the compiler to understand the intent of this object literal property, and not widen.

This would certainly be handy, but on its own it's still not ideal because one ends up writing `readonly` in a _lot_ of places. Take for example CSS-in-JS objects, which involve many properties whose types are unions of string literals. Every single one of those properties needs to then be marked `readonly` in order to defeat the scourge of type widening. But if one could declare an _entire object literal_ to have only read-only properties like so:

```ts
const o = readonly { x: 3, y: 'hello' };
// o: { readonly x: 3; readonly y: 'hello' }
```

now this starts to look like a workable solution! So, let this be my amended suggestion: `readonly` modifiers for both object literal properties as well as object literals themselves, which has the side effect of disabling type widening.

To the argument that `readonly` and type-widening are separate concerns and should be treated independently, to some extent I agree, but I also feel they are related; this is why the `let` vs. `const` rules for type widening exist, and I think that rationale makes sense. I also personally don't mind conflating `readonly` and type narrowing, because I prefer immutable objects everywhere too, so a syntax like

```ts
const o = readonly { x: 3, y: 'hello' };
```

would be attractive because it kills 2 birds with one stone. I currently don't use the `readonly` keyword much (even though in spirit I want everything to be read-only) because the cost-benefit ratio of annotating every single property in every single object literal is too high.

### LINKED ISSUE #26979: Type literal assertion `as const` to prevent type widening
## Search Terms
String literal type and number literal type in an object literal. Type assertion to make the string literal or number literal also a type literal. Cast string or number to type literal. Make a property definition in object literal behave like constant for the type inference. Define a property as string literal or number literal in an object. Narow the type of a literal to its exact value.

## Background
When I declare a constant and assign a string literal or a number literal to it, type inference works differently than when I declare a variable in the same way. While the variable assumes a wide type, the constant settles for a much narrower *type literal*.
```typescript
let   a = 'foo'; // type of a is string
const b = 'foo'; // type of b is 'foo'
```
This behavior comes in handy in many scenarios when exact values are needed. The problem is that the compiler cannot be forced to the "type literal inference mode" outside of the `const` declaration. For example the only way of defining an object literal with type-literal properties is this:
```typescript
const o = {
  a: 42 as 42,
  b: 'foo' as 'foo',
  c: 'someReallyLongPropertyValue' as 'someReallyLongPropertyValue',
  d: Symbol('a') as... ahem, what?
};
// type, as well as value, of o is { a: 42, b: 'foo', c: 'some…' }
```
That is not only incredibly annoying, but also violates the principles of DRY code. In adition it's really annoying.

## Suggestion
Since `const` is already a reserved keyword and can't be used as a name for a type, adding a type assertion expresion `as const` would cause no harm. This expression would switch the compiler to the "constant inference mode" where it prefers type literals over wide types.

```typescript
// In objects
const o = {
  a: 42 as const, // type: 42
  b: 'foo' as const, // type: 'foo'
  c: 'someReallyLongPropertyValue' as const
  d: Symbol('a') as const // type: typeof o.d
}


// In generic function
let f = <T>(a: T) => ({ boxed: a });
let x = f(42 as const) // T is 42, typeof x is { boxed: 42 }
```

## Non-simple types
There are currently three competing proposals for the way `as const` would work with non-simple types. 

### Shallow literal assertion
The first one would stay true to its name and would treat the type as if it were assigned to a constant.
```typescript
// Type of a is exactly the same as the type of b, regardless of X
const a = X;
let b = X as const;
```
However, this would mean that all non-trivial expressions would stay the same as they were without the `as const`. Therefore I would argue it's less useful than the other two proposals.

### Tuple-friendly literal assertion
The second proposal differs in the way it treats array literals. While the shallow assertion treats all array literals as arrays, the tuple-friendly assertion treats them as tuples. Then it recursively propagates deeper untill it stops at a type that is neither a `string`, `number`, `boolean`, `symbol`, nor `Array`.
```typescript
let a = [1, 2] as const; // type: [1, 2]
let b = [ [1, 2], 'foo' ] as const; // type: [ [1, 2], 'foo' ]
let c = [ 1, { a: 1 } ] as const; // type: [ 1, { a: number } ]

// Furthermore new syntax could be invented for arrays of const
// but that is outside the scope of this proposal right now
let d = [ [1,2] as const, [3,4] as const, [5,6] as const ]; // type: Array< [1,2] | [3,4] | [5,6] >
let e = [ [1,2], [3,4], [5,6] ] as const[]; // type: Array< [1,2] | [3,4] | [5,6] >
```
This is probably the most useful proposal, as it solves both the problem described in **Use Cases** and the problems described in #11152.

### Deep literal assertion
The third proposal would recursively iterate even through object literals. I included it just for sake of completeness, but I don't think it could be any more useful than the tuple-friendly variant.
```typescript
let c = [ 1, { a: 1, b: [1, 2] } ] as const; // type: [ 1, { a: 1, b: [1, 2] } ]
```

## Use Cases
Say I'm using a library which takes a very long object of type `LibraryParams` of various parameters and input data. I don't know some of the data right away, I need to compute them in my program, so I'd like to create my object `myParams` which I would fill and then pass to the library.

Since some of the properties of `LibraryParams` are optional and I don't want to check for them or assert them every time I use them – I know I've set them, right? – I wouldn't set the type of `myParams` to `LibraryParams`. Rather I'd use a narrower type by simply declaring an object literal with the things I need.

However some of the properties need to be selected from a set of exact values and when I add them to `myParams`, they turn into a `string` or a `number` and render my object incompatible with `LibraryParams`.

There are some ways around it using the existing code, none of which are particularly good. I'll give some examples in the next section.

## Examples
***Imagine that all of these examples contain much longer programs.***
```typescript
// Type is too wide because of LibraryParams

const myParams: LibraryParams = {
  name: "Foobar",
  favouritePrime: 7,
  sports: [ 'chess' ]
}

if (myFunctions.lovesFootball()) {
  myParams.sports.push('football'); //sports is possibly undefined 🤷
}

Library.doThings(myParams);
```

```typescript
// Type is too wide because of literal type widening

const myParams = {
  name: "Foobar",
  favouritePrime: 7,
  sports: [ 'chess' ]
}

if (myFunctions.lovesFootball()) {
  myParams.sports.push('football');
}

Library.doThings(myParams); //number is not assignable to prime 🤷
```

```typescript
// Too many type assertions

const myParams = {
  name: "Foobar",
  favouritePrime: 7,
  sports: [ 'chess' ]
}

if (myFunctions.lovesFootball()) {
   // this looks even weirder when you do it for the 10th time 🤮
  (myParams.sports as string[]).push('football');
}

Library.doThings(myParams);
```

```typescript
// Too much searching for the correct types

const myParams = {
  name: "Foobar",
  favouritePrime: Library.ParamTypes.Primes.Seven, // 🤮
  sports: [ 'chess' ]
}

if (myFunctions.lovesFootball()) {
  myParams.sports.push('football');
}

Library.doThings(myParams);
```

```typescript
// Probably the best solution but definitely not very dry

const myParams = {
  name: "Foobar",
  favouritePrime: 7 as 7 // 😕
  sports: [ 'chess' ]
}

if (myFunctions.lovesFootball()) {
  myParams.sports.push('football');
}

Library.doThings(myParams);
```

```typescript
// Top tier 👌😉

const myParams = {
  name: "Foobar",
  favouritePrime: 7 as const // 🤩
  sports: [ 'chess' ]
}

if (myFunctions.lovesFootball()) {
  myParams.sports.push('football');
}

Library.doThings(myParams);
```

## Checklist

My suggestion meets these guidelines:
* [x] This wouldn't be a breaking change in existing TypeScript / JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. new expression-level syntax)

## Related
#14745 Bug that allows number literal types to be incremented
#20195 A different approach to the problem of type literals in objects, excluding generic functions
#11152 Use cases for `as const` in generic functions

### PR DESCRIPTION
**Pull request by ahejlsberg**
With this PR we introduce `const` assertions which favor immutability in the types inferred for literal expressions (inspired by suggestions in #10195, #20195, and #26979). A `const` assertion is simply a type assertion that uses the reserved word `const` as the type name:

```ts
let x = 10 as const;  // Type 10
let y = <const> [10, 20];  // Type readonly [10, 20]
let z = { text: "hello" } as const;  // Type { readonly text: "hello" }
````

A `const` assertion establishes a _const context_ in which

* string, numeric, and boolean literals have non-widening literals types (see #11126),
* array literals have read-only tuple types (see #29435), and
* object literals have read-only properties.

Const contexts do not otherwise affect types of expressions, and expressions in const contexts never have a contextual type.

An expression `x` is said to occur in a const context if `x` is

* the operand of a `const` assertion, or
* a parenthesized expression that occurs in a const context, or
* an element of an array literal that occurs in a const context, or
* the expression of a property assignment in an object literal that occurs in a const context, or
* a spread expression that occurs in a const context.

Note in particular that const contexts extend into nested array and object literals. For example, the declaration

```ts
let obj = { x: 10, y: [20, 30], z: { a: { b: 42 } } } as const;
```

corresponds to

```ts
let obj: {
  readonly x: 10;
  readonly y: readonly [20, 30];
  readonly z: {
    readonly a: {
      readonly b: 42;
    };
  };
};
```

A `const` assertion requires the operand to be a string, number, bigint, boolean, array, or object literal, optionally enclosed in one or more levels of parentheses. It is an error to apply a `const` assertion to expressions of other forms.

Fixes #10195.
Fixes #20195.
Fixes #26979.

### CODE REVIEW

**Comment by jack-williams** [comment]
Did you consider using `const` as a hook for enabling deep excess property checking for arbitrary union and intersection types?

**Comment by j-oliveras** [comment]
A question:
Will be possible to write:
```ts
let obj = {
    x: 10,
    z: { a: { b: 42 } } as const
};
```
to initialize a type like:
```ts
let obj: {
  x: number;
  z: {
    readonly a: {
      readonly b: 42;
    };
  };
};
```

If yes, probably you want to add a test like that.

**Comment by pelotom** [comment]
This is AWESOME. Thank you @ahejlsberg!

**Comment by ahejlsberg** [comment]
> Did you consider using const as a hook for enabling deep excess property checking for arbitrary union and intersection types?

I did not. Can you provide an example of what you mean exactly?

> Will be possible to write (object literal with `const` assertions in some property values)

Yes.

> If yes, probably you want to add a test like that.

Sure.

**Comment by jack-williams** [comment]
@ahejlsberg 

Currently neither of these give excess property errors:
```ts
const u: { x: number } | { y: string } = { x: 42, y: "helloworld" };
const i: { a: { x: number } & { y: string } } & { b: boolean } = { a: { x: 42, y: "y", z: "excess" }, b: true };
```
Though these EPC bugs are on the backlog to fix for all literals, there are 2 obstacles:
1. Adding it would be a breaking change (but a good one IMO).
2. Adding it might introduce a perf. regression. I believe EPC is done today in a cheap way that can quickly prune large union types. I think fixing the general case requires a full relation check; there is no free lunch to be had.

I'm suggesting that maybe these give errors by using const contexts as a marker to enable complete and correct EPC:
```ts
const u: { x: number } | { y: string } = { x: 42, y: "helloworld" } as const; // error 'y' is excess (or some better error)
const i: { a: { x: number } & { y: string } } & { b: boolean } = { a: { x: 42, y: "y", z: "excess" }, b: true } as const; // error 'z' is excess
```
Basically using const contexts as a way to trial full-EPC in a way that does not introduce any breaking changes. If the trial is successful then full-EPC can be rolled out to all literals with const being ahead of the curve; if the trail does not work out then full-EPC can be removed in const contexts in a non-breaking way.

**Comment by Meligy** [comment]
Can it be also assumed in certain situations?

Like:

```ts
const a: 12; // type is 12
let x = {
  // Can this be `const`? 
  // Either always, or only when it's `readonly x1: a`. It's fine either way.
  x1: a 
};
```

The situation here being assignment to a constant, and fine to require `readonly` for this.

Also, one important situation (as [mentioned with more details](https://github.com/Microsoft/TypeScript/issues/20195#issuecomment-455935695)):

```ts
function createAction<TType>(type: TType) {
    return {
        type: type
    };
}

const doStuff = createAction("DO_STUFF");
```

The situation here being sending a literal (string or number) to a function with argument type being generic and inferred from usage. It'll be nice to avoid `const` in `createAction("DO_STUFF" const)`.

**Comment by qm3ster** [comment]
Next step: allowing `import`ing JSON modules `as const`

**Comment by tal** [comment]
Is the `const` keyword from somewhere else? To me it's confusing and easy to conflate with the `const` variable declaration keyword. `immutable` would be clearer for me.

**Comment by pelotom** [comment]
> Is the `const` keyword from somewhere else? To me it's confusing and easy to conflate with the `const` variable declaration keyword. `immutable` would be clearer for me.

That would require adding a new keyword for dubious benefit. Personally I think `readonly` would have been preferable to `const` (since it has the effect of adding `readonly` modifiers deeply through a tree of types). But `const` has the advantage of being short 🤷‍♂️

**Comment by pelotom** [comment]
> Maybe `literal` would do it justice?
> 
> ```ts
> let x = 10 as literal; 
> const y = 10 as literal; 
> ```

But it was already a literal, before you added `as literal`...

**Comment by stnwk** [comment]
> > Maybe `literal` would do it justice?
> > ```ts
> > let x = 10 as literal; 
> > const y = 10 as literal; 
> > ```
> 
> But it was already a literal, before you added `as literal`...

Haha true! Thanks for clearing that up, edited my reply.

**Comment by aleksey-bykov** [comment]
be it a `wanker` i would not say a thing, but sure let's nitpick and bike shed this thing to the bone

**Comment by ducin** [comment]
Question: if the `as const` clause recursively adds `readonly` to object literals, why not calling it `as readonly`? Not a big issue, but introduces some ambiguity with the native ES6 `const` which has slightly different meaning than TS' `readonly`

**Comment by aleksey-bykov** [comment]
to those who believes in `readonly` modifiers, will it make your array properties readonly "deeply through a tree of types"?

**Comment by joshburgess** [comment]
@ahejlsberg Would there be anyway to dynamically concatenate two literal types to produce a new literal type using this `const` context?

For example, in a function like...

```ts
const concatLiterals = <A extends string, B extends string>(a: A, b: B) =>
  `${a}{b}` as const // or (a + b) as const
```

I'm assuming the dynamic nature of the above here would prevent this from working, and it would still widen the type to `string`, but this capability is something I've wanted for a while.

It would be great to have some sort of type-level concat helper for literals like this.... so that you could do something like:

```ts
(a + b) as const<A + B>
```

or

```ts
(a + b) as const<Concat<A, B>>
```

or something along those lines.

**Comment by pelotom** [comment]
Apparently `as readonly` instead of `as const` was considered and decided against, see the meeting notes: https://github.com/Microsoft/TypeScript/issues/29568. I understand the reasoning, but I still think `as readonly` is more intuitive 😛

**Comment by DanielRosenwasser** [comment]
I agree with `as const` being confusing. Maybe the only reason I don't feel as strongly against it is the explanation itself is relatively consistent with the current distinction of `const` and `readonly`.

I suspect that we'll end up with a quick fix like `Did you mean 'as const'?` when someone writes `as readonly`.

**Comment by aleksey-bykov** [comment]
oh let's not forget what readonly really is https://github.com/Microsoft/TypeScript/issues/13002

**Comment by DanielRosenwasser** [comment]
Somehow I knew exactly what you were linking to before I clicked on it.

**Comment by Bnaya** [comment]
This is probably out of the scope, but:
it will be nice to have a way to also "kill" methods that mutating objects in place.
Such as: Array.sort, Array.splice, Date.setDate

**Comment by elektronik2k5** [comment]
+1 for `as readonly` because that's what it actually does.
TypeScript already has a high syntax overhead and adding more cognitive load by requiring the user to keep in mind how `readonly thing` translates and relates to `thing as const` should be avoided.

**Comment by masterkidan** [comment]
A question regarding this.... 
One of the usages of the Readonly type is to allow methods to return a Readonly object like so 
```ts
interface foo {
   bar: string
}
function foobar(): ReadonlyArray<foo> {
}
```

Is it possible to return something like so 
```ts
function foobar(): foo as const {
}
```
Would really like to have  a use case like the above included as well, as it would help a lot with immutability.

**Comment by aleksey-bykov** [comment]
`as readonly` has very little to do with the "true" readonly (think of readonly arrays)
more to that, the "true" readonly is an unreachable goal given the complexity it takes to make it really "true"
so please calm down and leave it as `as const` because this is more accurate name to the job that this modifier is doing: it doesn't make (or even try to make) your objects readonly, the only thing it states is that properties of your object *are of the exact type as written* (which makes it const, according to how the `const` currently works)

**Comment by masterkidan** [comment]
@aleksey-bykov : Does that mean I can reassign properties within an object that is a const? meaning can I do this 
```ts
let a = { b: 42 } as const;
a.b++; // Allowed
a = { b: 43 }// Also allowed?
```

**Comment by aleksey-bykov** [comment]
`a.b ++;` must be disallowed on `let a = { b: 42 } as const;`, so the answer is no, you cannot do it, because there is no such thing as:
 - incrementing on `42` and expecting it to be of type `42`

**Comment by aleksey-bykov** [comment]
i am saying that once this pull request is merged, `a.b` can only be assigned by a value that is proven to be `42 ` at design time, currently in TypeScript there is only one way to prove it: 
 - by **literally writing** `42` and **nothing else** in your code
 - by assigning from something that has been proven to be of type `42`

not sure if this explanation helps, bottom line is, unless it's of type `42` it cannot be near `a.b`

**Comment by aleksey-bykov** [comment]
here is an illustrative example about the difference between `readonly` and `const`, @ahejlsberg please correct me if i am wrong with your idea of `as const`
```
const obj = { x: 1 } as const;
obj.x = 1; // ALLOWED!!!
```
vs.
```
interface Obj { readonly x: 1 }
const obj: Obj = { x: 1 };
obj.x = 1; // DISALLOWED!!!
```

**Comment by jack-williams** [comment]
I think that `Obj` is the very type inferred by  `as const` in your first example.

**Comment by qm3ster** [comment]
@aleksey-bykov no, the two examples would be the same, `DISALLOWED` as you say.
In both, `typeof obj` would be `{ readonly x: 1 }`, as written in original description.
(as opposed to `{ x: 1}`, still a very distinct type from `{x: number}`)

So, the `as const` has two distinct effects: it types the insides as unique literals, recursively, same as 
```ts
let b = 2 // b: number
const a = 1 // a: 1
```
But also recursively applies readonly to all layers.

It's helpful because we go from this:
```ts
const a = 1
interface B<T extends number> { a: T }
const makeB = <T extends number>(num: T) => ({a} as B<T>)
const b = makeB(a)
const c = b.a
b.a = a // pointless, why did you write this?
b.a = 1 // pointless again.
b.a++ // NOT an error! HAHA SILLY TYPESCRIPT WAT IS U DOING
b.a == 1 //true
b.a--
b.a = 0 // Finally ERROR
```
to this:
```ts
const a = 1
interface B<T extends number> { readonly a: T }
const makeB = <T extends number>(num: T) => ({a} as B<T>)
const b = makeB(a)
const c = b.a
b.a = a // ERROR, no doing dumb stuff
b.a = 1 // ERROR
b.a++ // ERROR, the day is saved
```

**Comment by aleksey-bykov** [comment]
even better this way, if you are right (haven't tried this branch yet)

**Comment by ahejlsberg** [comment]
Some rationale on `xxx as const` vs `xxx as readonly`.

* `const` is shorter than `readonly`. Shorter is better.

* It is not possible to declare a type named `const` because `const` is a reserved word. Not so with `readonly`. Therefore, using `readonly` would be a breaking change.

* One of the effects of const contexts is to use non-widening literal types for string, number, and boolean literals. `10 as const` does a fair job of indicating that. `10 as readonly` does not. It would be pretty much meaningless (literals are inherently readonly so what would it mean to make them more readonly?).

* `readonly` doesn't imply immutability. `readonly` means that _you_ can't modify it, but someone else might be able to by maintaining a non-readonly alias to the same location. `const` on the other hand _does_ imply immutability. `const` means you can't modify it, _nor can anyone else_. So, `const` is a stronger contract than `readonly`. Now, in a const context, properties of object literals and elements of array literals are not just `readonly`, they're also `const`. Why? Because references to the objects allocated by the literals have never been revealed to anyone else. Thus, they couldn't have been aliased and therefore no one else can modify them (unless they break type safety, but that's beside the point).

For these reasons we feel `const` is the better choice.

**Comment by aleclarson** [comment]
Quick aside: Will `Readonly<[1, 2]>` be changed to produce `readonly [1, 2]` once `as const` lands? Otherwise, it seems you must do `const tup = [1, 2] as const` with `type T = typeof tup` to get a readonly tuple type.

**Comment by ahejlsberg** [comment]
> Will `Readonly<[1, 2]>` be changed to produce `readonly [1, 2]` once `as const` lands?

Yes, that's already happening in #29435, a prerequisite for this PR.

**Comment by karol-majewski** [comment]
Great stuff. The community [seems to be very excited about it](https://twitter.com/WrocTypeScript/status/1088431175025283073).

These two concerns were expressed in the comments:

- possibly confusing syntax (`as immutable` is the most popular suggestion so far)
- how does one update such a structure?

**Comment by pelotom** [comment]
> * how does one update such a structure?

One doesn’t. That’s what immutable means.

**Comment by ilyaztsv** [comment]
agree with 

> possibly confusing syntax (`as immutable` is the most popular suggestion so far)

**Comment by qm3ster** [comment]
> agree with
> 
> > possibly confusing syntax (`as immutable` is the most popular suggestion so far)

:thinking: maybe `as frozen` :rofl:  
I get why `const` is not breaking at all since it's a reserved word, but do we really need to care that much? About a lowercase type?

To me `as const` sounds like "as if it was assigned to `const` on creation".
Which in the current case, sadly is unique types for primitive literals but widened types for objects.

Does it really make sense that 
```ts
const immutable = {/* expression */}
// is a different type than
{/* expression */} as const
```

**Comment by vmasek** [comment]
Will it be possible to add a compiler option that would automatically enable this for all `const` and also maybe `readonly`? 

Our team already tries to not use mutable operations and have them disabled by tslint-immutable. With a compiler option, it would be great to have immutability mandatory for constants.

**Comment by tycho01** [comment]
@vmasek:
> Will it be possible to add a compiler option that would automatically enable this for all `const` and also maybe `readonly`?
> 
> Our team already tries to not use mutable operations and have them disabled by tslint-immutable. With a compiler option, it would be great to have immutability mandatory for constants.

I definitely hope they feel more bullish on this than [last year](https://github.com/Microsoft/TypeScript/pull/17785#issuecomment-383639234)!

**Comment by MastroLindus** [comment]
I am really excited for this change, I have been waiting for something like it for a long time, thanks for delivering it.

I am adding a +1 on the request to have a flag to enable this as a default behavior for const assignments.
The rationale is that people that are going to use this feature are probably willing to use it as much as possible, basically everywhere, and having to manually add the "as const" every time would be both error-prone and cumbersome (but still much better than any other workaround we had so far to achieve something similar)

**Comment by RyanCavanaugh** [comment]
PR targets the wrong branch - I believe you'll have to open a new one. Or you can merge this and then separately merge the feature branch into master

**Review by RyanCavanaugh — APPROVED** [review]

**Comment by aleksey-bykov** [comment]
@ahejlsberg got my code broken after tonight's update, not sure if it's directly related: https://github.com/Microsoft/TypeScript/issues/29662

**Comment by tjfryan** [comment]
Will it be possible to combine this feature along with conditional types (as demonstrated in https://github.com/Microsoft/TypeScript/issues/17915#issuecomment-413347828) to refine a discriminated union to more than one (but not all) of the possible types? For example if I were to try a similar example to that link where an array of tags was accepted, as long as the tags were specified statically in my codebase, could I do this?

```typescript
function cast<A extends Foo['tag']>(foo: Foo, ...tags: A[] as const): FindByTags<Union, typeof tags> {
  ...
}
```

**Comment by qm3ster** [comment]
Since this was merged, can we maybe get something related for importing JSON modules `as const`?

**Comment by xAlien95** [comment]
I'm currently using `typescript@next` and I'm finding this feature really useful. There's only a behavior that I don't know if it's expected or not.

We can define an array of arrays `as const` and correctly get the types of every child array and child array element:

```ts
const arrayOfArrays = [
  ['One', 'Two', 3],
  ['Three', 'Four'],
] as const;

type T1 = typeof arrayOfArrays[0][number]; // T1 = 'One' | 'Two' | 3
type T2 = typeof arrayOfArrays[1][number]; // T2 = 'Three' | 'Four'
type T3 = typeof arrayOfArrays[number][0]; // T3 = 'One' | 'Three' <-- amazing!
type T4 = typeof arrayOfArrays[number][number]; // T4 = 'One' | 'Two' | 3 | 'Three' | 'Four'
```

However, using an array of objects `as const`, we cannot get all the literal types unless we define an intermediate variable:

```ts
const arrayOfObjects = [
  {
    values: ['One', 'Two'],
  },
  {
    values: ['Three', 'Four'],
  },
] as const;

type T1 = typeof arrayOfObjects[0].values[number];
// not valid
// - Block-scoped variable 'values' used before its declaration. ts(2448)
// - Variable 'values' is used before being assigned. ts(2454)
// - 'number' only refers to a type, but is being used as a value here. ts(2693)

type T2 = typeof arrayOfObjects[0].values;
// not valid
// - Block-scoped variable 'values' used before its declaration. ts(2448)
// - Variable 'values' is used before being assigned. ts(2454)

type T3 = typeof arrayOfObjects[0];
// valid
// T3 = { readonly values: readonly ["One", "Two"] }

// workaround

const values0 = arrayOfObjects[0].values;
type T4 = typeof values0[number];
// T4 = 'One' | 'Two'
```

Is this the expected behavior? Shouldn't a "literal" list of object return literal types for each specified index and union literal types for multiple indexes?

**Comment by dragomirtitian** [comment]
@xAlien95  wrong syntax. This is the same syntax as before 3.4

```ts
const arrayOfObjects = [
  {
    values: ['One', 'Two'],
  },
  {
    values: ['Three', 'Four'],
  },
] as const;

type T1 = typeof arrayOfObjects[0]['values'][number];
```

**Comment by maartenth** [comment]
This is great! would it also be possible in the future to import a json file `as const`? A json file contains an object literal, but you cannot mark it `as const` in the json file itself...

**Comment by maartenth** [comment]
Now I see that @qm3ster already asked this on Feb 15...
> Since this was merged, can we maybe get something related for importing JSON modules `as const`?

It would be a great addition, because these are now the only object literals you cannot assert as const. The `as const` syntax may conflict with the existing `import as` syntax but the good news is that the angle bracket `<const>` syntax does not conflict with jsx in import statements.

Like:

```ts
import <const>data from 'data.json'
```

**Comment by qm3ster** [comment]
@maartenth I was thinking 
```ts
import as const * as data from 'data.json'
//or
import data from 'data.json' as const //my preference, and generally the position in which I would like to cast imports, if that feature ever becomes available.
```

**Comment by scvnathan** [comment]
I'm not sure if this is intentional, but I've noticed that when constraining to a string, TS can narrow to the string literal.
For example, in @Meligy's example:
```
function createAction<TType>(type: TType) {
    // Not concerned with return types here, but inferred argument type
    return {
        type: type
    };
}

// Today this gets inferred as string, but can string literal be inferred as itself?
const doStuff = createAction("DO_STUFF");
```

Change the signature to `function createAction<TType extends string>(type: TType)` and `TType` gets inferred to "DO_STUFF"

**Comment by Meligy** [comment]
This is awesome. I'm surprised I haven't tried this before or thought it wouldn't work.

If you want to support all types, but be restricted about them (infer the exact literal if it's a string or number or enum), you can do this:


```ts
function createAction<TType extends string | {} | null | undefined>(type: TType) {
    // Not concerned with return types here, but inferred argument type
    return {
        type: type
    };
}
```

Test cases:

```ts
const doStuff = createAction("DO_STUFF");

const doStuff2 = createAction(1);

const doStuff3 = createAction(null);

const doStuff4 = createAction(undefined);

enum TestEnum {One, Two}
const doStuff5 = createAction(TestEnum.One);
```

Seeing it in [unofficial TypeScript playground][1]

[1]: https://typescript-play.js.org/?ts=2.4.1#code/GYVwdgxgLglg9mABBATgUwIZTQQWvMAHgBViBPABzUTQA9swATAZ0WahRjAHNEAfRAG8Avv0RgQAG0ljwjNMC5pGAPgAUUSmgBciUloCUQgFCIziAPQXEAOThRkCCGhRhliAO4woAC0TooEFdETSpWHxc0ABpEACMQBy5gF3RGRAwUbhAAWzQwB1C0U3MAoKRBYvNzQt1CysRhAG5jYWNjCAR2REY4AGVA4GBEAF5kdCxcfAQ1ACIAEQB5AH1e4gBVADENmYNm9s6HHv6QQYAmEbHMbDxYaYBGXbaOsC6jgeAAZgvUK8nbsDUEmkj32L0OfXeABZvuNrlMAXIFEpGCC8jk9Gh2ABRCTZIQLNwxYgeOCtZ6vCEnYAAVhhvxuBDUxExUBxOQAdAS0LsgA

**Comment by MuYunyun** [comment]
Does [@babel/plugin-transform-typescript](https://babeljs.io/docs/en/next/babel-plugin-transform-typescript.html) support this feature —— Const contexts for literal expressions?

**Comment by pelotom** [comment]
@MuYunyun yes... I suspect because it was already supporting any `x as foo` expression by just erasing the `as foo` part.

**Comment by MuYunyun** [comment]
@pelotom any x includes ’Reserved words‘?

**Comment by pelotom** [comment]
@MuYunyun you mean does `foo` include reserved words? I'm not sure, all I know is Babel seems to transpile `as const` just fine.

**Comment by brandonkal** [comment]
I am glad to see this in! I was ready to suggest `as self` or something along those lines, because it is essentially the same as doing as repeating the declaration:
```js
const theme = {
  big: '10em',
  small: '5em'
} as {
  big: '10em',
  small: '5em'
}
```
Using a JS keyword feels a little odd but I'm glad to type less regardless.

**Comment by rjamesnw** [comment]
Is there a way to allow constant string to represent string literals?  For example:

    const moduleName = "../../../Project/src/Mod";

    async function doSomething() {
        var mod = await import(moduleName); // (works only if a string literal at the moment)
    }

In the case above, "mod" is of type "any" because the compiler doesn't recognize the string literal in the constant `moduleName`.  I'm not sure if this was an oversight, but it makes sense to allow it, since constant strings cannot be reassigned.

**Comment by Meligy** [comment]
@rjamesnw since this PR is already merged I don't expect that you can get traction for anything new, but this is a good one actually. Can you please open a new issue for it? Thanks.

**Comment by rjamesnw** [comment]
You’re right, sorry, it also occurred to me and I was going to get to doing so asap. ;)

**Comment by AliN11** [comment]
When someone would use `let` for declaring a variable which is typed with `as const`?

**Comment by BeautyInNorth** [comment]
Excuse me,why is the error reported?
```
const arr = [
  { age: 12, fruits: 'apple' },
  { age: 16, fruits: 'cherry' },
  { fruits: 'banana' },
] as const

type Fruits = typeof arr[number]['fruits']

arr.map(({ age, fruits }) => {
  //Property 'age' does not exist on type 
  //'{ readonly age: 12; readonly fruits: "apple"; } | { readonly age: 16; readonly fruits: "cherry"; } | { readonly fruits: "banana"; }'.
  console.log(age)
})
```

**Comment by pelotom** [comment]
> Excuse me,why is the error reported?

Because not every element of the array has an `age` property.

**Comment by AliN11** [comment]
I found out that `as const` does not work after a variable has been declared:
```ts
let x = 12;
x = 2 as const;

x = 5;

console.log(x); // 5
```

**Comment by brandonpittman** [comment]
> > Excuse me,why is the error reported?
> 
> Because not every element of the array has an `age` property.

Is there a way to type an array of objects declared using `as const` if some of the objects lack properties that other objects have?

**Comment by m93a** [comment]
image

_While `as const` turned out to be a really nice feature, it does have its drawbacks._
